crypto/asn1/a_time.c

```diff
@@ -98,7 +96,93 @@ ASN1_GENERALIZEDTIME *ASN1_TIME_to_generalizedtime(ASN1_TIME *t,
 
     OPENSSL_strlcat(str, (char *)t->data, newlen);
 
+ done:
+    if (out != NULL && *out == NULL)
+        *out = ret;
+
+    return ret;
+ err:
+    if (out == NULL || *out != ret)
+        ASN1_GENERALIZEDTIME_free(ret);
+    return NULL;
+}
+
+/* Convert an ASN1_TIME structure to GeneralizedTime */
+ASN1_UTCTIME *ASN1_TIME_to_utctime(ASN1_TIME *t, ASN1_UTCTIME **out)
+{
+    ASN1_UTCTIME *ret;
+    int newlen;
+    char newstr[20];
+
+    if (!ASN1_TIME_check(t))
+        return NULL;
+
+    if (out == NULL || *out == NULL) {
+        if ((ret = ASN1_UTCTIME_new()) == NULL)
+            goto err;
+    } else
+        ret = *out;
+
+    /* If already UTCTime just copy across */
+    if (t->type == V_ASN1_UTCTIME) {
+        if (!ASN1_STRING_set(ret, t->data, t->length))
+            goto err;
+        goto done;
+    }
+
+    /* check the year 1950..2049 */
+    if (t->data[0] == '1') {
+        if (t->data[1] != '9' || t->data[2] < '5')
+            goto err;
+    } else if (t->data[0] == '2') {
+        if (t->data[1] != '0' || t->data[2] > '4')
+            goto err;
+    } else
+        goto err;
+
+    /* copy YYMMDDhh - ignore century */
+    memcpy(newstr, &t->data[2], 8);
+    newlen = 8;
+
+    /* mm? Optional on GENERALIZEDTIME */
```

Comment:
I've seen conflicting sources as to whether minutes are optional in GENERALIZEDTIME.